### ISSUE TITLE
Make Core GPU Buffers Re-Sizable

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#38]
To make things easy, I initially set up the core buffers to have a fixed size. As I have added more data buffers though, this has led to a weird mismatch of max size choices. There's really not a good upper limit that can be set on each individual buffer, so i have just chosen from one for two maximums, points or hulls. Any buffer that is defined has one of these two values set for the maximum.  

Generally speaking, GPU kernels need to have fixed size memory buffers, which is why these have limits in the first place. That said, this doesn't mean the buffers have to be fixed _forever_, they just need to have a defined maximum when being used. Between frames, there is nothing that says a buffer can't be resized. And this is exactly what needs to occur here.

The current maximum for each buffer should be kept in memory, and when data is added, it is checked to see if the new data would exceed the max. If so, a separate buffer can be created that is larger, and the old buffer data can simply be copied into the new, larger buffer. The old buffer can then be destroyed.

Of course, some practical initial limit should be set, as well as a "growth factor" that determines how much the buffer will grow when it is found to be too small. It is probably easy enough to simply double the size, with some upper limit beyond which instead of doubling, it grows by some fixed increment. 

It will also be beneficial to be able to reduce the buffer size as items are removed, for example when objects are deleted from the game world, so whatever buffer sizing code is created it should be able to account for both operations, though reducing size should probably be done somewhat lazily, during a times of inactivity. As I generally expect the game world to be open ended, I don't think there will be natural transition points, but if possible the design should probably expose some kind of "down time" call that can be used as a signal to compact things.

**Comment by controllerface** [comment]
So, after taking some time to reason through things, one really big hurdle for this work is going to be making sure kernels are always updated when buffers change size. At the moment, memory is initialized up front, and then kernel args are set. However, if buffers need to change size, I don't think this will work anymore, as they will need to be reset as they change.

Because of this, I will need to have some way of programmatically storing which kernels need to be updated so that they can be, every time a buffer changes size. It would be nice if there was a way to keep the same pointer to a buffer when changing it, but I seriously doubt that will be possible. I'll do a quick check, _just in case_ I can find anything that allows this, but I'm not going to get my hopes up.

I may need to come up with a functional mechanism that can be stored within the buffers, and then when they resize, loop through the list and call the function with the new pointer. I could also have some list or map that holds the kernel and argument position, and loop through that. Otherwise it would be a seriously janky mess of manual update blocks, which I am not even going to consider.

**Comment by controllerface** [comment]
Oh and another thing I will need is some clear way to call an "ensure size" before each use of the buffer. Not all kernels add data to buffers, so this will probably need to be a case-by-case thing vs something that every kernel will need. Mainly the GPU-CRUD kernels.. actually possibly only those. It may be worth splitting those and their buffers into a separate class.

The more I dig, the more I think it may be wise to do the GPU class refactor before attempting this, as it will probably make certain things more apparent since that will require determining where the API boundaries are.

**Comment by controllerface** [comment]
I think I will need to come Up with a "Resizable Buffer" class that can be reused in several classes and that Will support Some kind of "Registration" mechanism, at I mentioned before. However , I think I will possibly start moving Some buffers into the "Core" Memory Class, and Keep some inside the GPU class. Originally, I wanted to keep things centralized, but I think it may be beneficial to have buffers associated with classes that provide relevant Functionality. 

To make this work with kernels, I will probably want to make a new method that accepts the new buffer type, and does the registration step automatically. The hope is that there won't need to be a lot of extra work to make the resizing buffers work correctly.

**Comment by controllerface** [comment]
In order to make buffer resizable, at least two key factors will need to be tracked:

1. The current capacity of the buffer. This is essential to ensuring that subsequent calls to put data into the buffer will succeed. Before data can be inserted into the buffer, this value will need to be increased to accommodate unless there is enough space.
2. The current _used_ portion of the buffer. When a buffer grows in size, it will probably not only grow enough to accommodate one more item, but a number of items. In addition to ensuring the buffer itself is big enough when inserting data, this will be checked to ensure the buffer has enough space to add items. 

All code paths where data is added to a buffer will have to be updated with some "ensure size" logic that calls a method which does just that. It will need to check both the capacity and the used space and if neither is sufficient the buffer will need to be expanded. 

It will probably be wise to also include some mechanism for shrinking the buffer if it falls below some threshold, so that memory does not sit unused, but the first step will be just to ensure that we can grow as needed.

**Comment by controllerface** [comment]
I have started to go through the current buffer enum and find buffers that can be isolated and moved under other classes as private members. This should give me a way to experiment a bit with building out an interface for resizing buffers with less impact.

It looks like there are only really a handful of buffers that can be isolated this way, but that should still be good enough to get things rolling. It looks like the physics reaction buffers and the compaction "shift buffers" are really the only ones that stand out. 

As before, going through this makes exercise makes me wonder if I should block this work again on a different task, this time the debug UI, so that these buffer values could be more easily read quickly, on a dashboard.

Will think on this...

**Comment by controllerface** [comment]
I decided to go for it, and I have the first resizable buffer implementation, which is for buffers that don't need to preserve data between frames. This seemed like the simpler use case, and it allowed me to actually improve longer term efficiency (at least on paper) in a few places at the expense of a bit of "buffer churn" when the app starts up. Basically, there were several buffers that were created and destroyed every frame, that now will just grow an be reused instead. This means they will have a tiny bit of extra overhead at first, since now there is a check to see if the buffer is big enough before it is discarded and replaced with a bigger buffer, but then after the buffer grows enough to accommodate everything, only the check is performed instead of the buffer creation. In the grand scheme of things, a single arithmetic check is a lot cheaper that destroying and recreating a buffer.

In doing this work, I did discover a bug with intellij, which causes it to crash or lock up after several repeated test runs. Being a bit goofy, i used a stream in the first implementation attempt of the resizing buffer and this made the problem a lot easier to trigger, which in the end was actually a good thing, since I am more aware of the issue. I would not use a stream in the critical path normally, as it's obviously slower than a basic for loop, but helping to highlight an issue like this is certainly a bonus. I ran many tests outside of the IDE from a JAR and could not replicate it, so I am at least relieved it isn't something in my code. The resizing buffer approach certainly is going to have some extra overhead in some cases, but the tradeoff of not needing to decide how many of each core object must be made at the beginning of the program is kind of necessary in the long term, such a drawback would never be acceptable in an actual game.

In any case, I am happy with how this is going, the next steps will definitely require careful thought. For now I'm happy I was able to successfully move the physics and shift buffers inside the classes that use them, taking more responsibility away from the central GPU class. It is already easier to understand what is going on with these things all in one place, and I'm hopeful this trend continues in the next set of updates.

**Comment by controllerface** [comment]
Took a small detour to refactor the edge buffer a bit, as it was causing some issues. This buffer was among the very first I created when starting this project, and it was from a point in time where I didn't full grasp how Open CL worked as well as I do now. As such, I opted to cram an `int2`, a `float`, and an `int` buffer all into a single `float4` buffer, casting out the `int` and `int2` when needed. As it turns out, this made for some bizarre rounding issues when moving data between buffers, and since the integer values were used to index into other buffers, this broke things in interesting ways.

I also found, after many crashes and a LOT of experimentation, that there is a limit to how rapidly you can get away with resizing a buffer. I initially chose a rather modest increase at each point where the buffer needed to grow, and this simply did not cut it. After a lot of headache, I finally realized that the code was working correctly on paper, but was locking up just due to the sheer number of buffer resizes that were happening. In the end, I went with a small-ish initial buffer size, but then when increasing, the growth factor is significantly larger. 

Rather than use a specific size in bytes, KB, etc., I opted for an "item capacity" increase. So buffers of say, `float4` will increase when it runs out of some number of "slots" in the buffer, and that number is the same for an `int` buffer (or any other buffer type) so they grow by the data width * growth factor. In the end, I am still using significantly less overall space per buffer than I was before, so I think it's a win. 

Now, the real work can begin.. time to move ALL buffers over to the new re-sizing scheme and _hopefully_ observe no outward impact. Visually, everything should look the same in terms of framerate and stability. Though I do hope I can push the number of squares and triangles in the single object tests up a bit higher than they can now. Because the amount of memory they can use is hard-capped, I don't really know how many I can spawn yet.

Oh and it occurs to me that I need to open a ticket (if I don't get to it in this work) to add some kind of hard limit on the number of collisions that can occur every frame. I have probably had to restart my machine more times in the last few days than in all the past year combined just because I kept spawning too many objects on top of each other.

**Comment by controllerface** [comment]
All buffers have been converted, and now the _largest_ test I can run tops out at just over 25 MB of buffer usage. This is dramatically lower than before, where the buffer total was over 500 MB. And that did not even count a lot of the transient buffers that are now being captured in the logging. The smaller tests, which I tend to run a lot more often run at just a little over 9 MB. 

It took a bit of finagling to make sure the performance was the same, and I learned _a lot_ about how to make this all work in real-time. Really happy now though, since I am not worried about hitting a memory cap. I do have to be careful with some things, like the rate of adding new objects, as well as the rate of buffer resizing, and doing the scans for deleting them as well. But in the end, the code is now much better positioned to behave like a real game would. Still a ways to go before it's actually at that stage, but this is a very big milestone in my book.

In addition to adding some kind of collision ceiling, I also think I will need to do something similar for object deletion. I have noticed that the system can get unstable when a large number of objects get deleted at once, as well as potentially locking up, and it's not entirely repeatable. I suspect that I probably could get some extra safe guards by adding more event waits, but I really want to avoid that if possible because it really eats into frame time. I want to continue to target 60 FPS.

Anyway, PR incoming. Glad to put this one to bed!
